### DISCUSSION TITLE
Problem with calculating %ROPE for 2 samples t-test

### DISCUSSION THREAD

**Discussion by sloughon** [easystats/bayestestR#495] [Q&A]
Hello,

First of all, I would like to apologize for the triviality of my question, but I am just getting started with the Bayesian approach and this is my first analysis in R.

I was trying to calculate the proportion of samples inside the ROPE. My data is two vectors, one containing the results of group 1, the other containing the results of group 2 (independent samples):

```
library(BayesFactor)
library(bayestestR)
a=c(value1, value2...)
b=c(value1, value2...)
result <- ttestBF(a,b)
rope(result, range = c(-0.5, inf), ci = 1, ci_method = "HDI", verbose = TRUE)
```
It gets the correct BF values, however, the % in ROPE are a) do not correspond in any way to the expected value b) are different each time the code is run. I guess I'm making a very basic mistake. 

Additionally, I would like to make sure - in such a situation, the ROPE range relates to the size of the effect? I plan to use %ROPE as a measure of how likely an effect that is no greater than small is.

I will be grateful for any help.

**Comment by bwiernik** [comment]
Can you give a specific real example? The exact values are expected to change slightly from run to run because they are based on simulated values.

**Comment by mattansb** [comment]
All the `bayestestR` functions (via `insight::get_parameters()`) use the **raw** difference.

If you want the standardized difference (aka Cohen's *d*), you can:

``` r
d <- 0.15
n <- 100

a <- 10 + rnorm(n)  * 30
b <- 10 + rnorm(n, mean = -d) * 30


suppressPackageStartupMessages(library(BayesFactor))
library(bayestestR)

result <- ttestBF(a, b)

result_posts <- posterior(result, iterations = 4e3)
result_posts <- as.data.frame(result_posts)

d_posts <- result_posts[, "delta", drop = FALSE] # delta is the std diff

rope(d_posts, range = c(-0.2, 0.2))
#> # Proportion of samples inside the ROPE [-0.20, 0.20]:
#> 
#> Parameter | inside ROPE
#> -----------------------
#> delta     |     83.98 %
```

<sup>Created on 2021-12-21 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

---

@strengejacke Does it make sense for `insight` to give only the raw difference?

**Comment by strengejacke** [comment]
I thought that for `BFBayesFactor` objects, `get_parameters()` returns draws from the posteriors of that object. Isn't this the most straightforward way?

**Reply by mattansb** [reply]
`insight::get_parameters()` only gives the raw difference:

``` r
suppressPackageStartupMessages(library(BayesFactor))
library(bayestestR)

a <- sleep$extra[sleep$group == 1]*3
b <- sleep$extra[sleep$group == 2]*3

BF_ttest_ind <- ttestBF(a, b)
BF_ttest_paired <- ttestBF(a, b, paired = TRUE)



insight::get_parameters(BF_ttest_ind) |> describe_posterior()
#> Summary of Posterior Distribution
#> 
#> Parameter  | Median |        95% CI |     pd |          ROPE | % in ROPE
#> ------------------------------------------------------------------------
#> Difference |  -3.42 | [-8.57, 1.04] | 93.20% | [-0.10, 0.10] |     0.84%

# beta - raw diff (insight::get_parameters)
# delta - std diff
posterior(BF_ttest_ind, iterations = 4000) |> as.data.frame() |> 
  describe_posterior()
#> Summary of Posterior Distribution
#> 
#> Parameter    | Median |         95% CI |     pd |          ROPE | % in ROPE
#> ---------------------------------------------------------------------------
#> mu           |   4.62 | [ 1.97,  7.21] | 99.88% | [-0.10, 0.10] |        0%
#> beta (x - y) |  -3.41 | [-8.85,  1.13] | 92.67% | [-0.10, 0.10] |     1.29%
#> sig2         |  33.52 | [14.95, 61.13] |   100% | [-0.10, 0.10] |        0%
#> delta        |  -0.59 | [-1.48,  0.21] | 92.67% | [-0.10, 0.10] |     7.92%
#> g            |   0.68 | [ 0.04, 10.12] |   100% | [-0.10, 0.10] |     2.76%


insight::get_parameters(BF_ttest_paired) |> describe_posterior()
#> Summary of Posterior Distribution
#> 
#> Parameter  | Median |         95% CI |     pd |          ROPE | % in ROPE
#> -------------------------------------------------------------------------
#> Difference |  -4.29 | [-6.66, -1.47] | 99.83% | [-0.10, 0.10] |        0%


# mu - raw diff (insight::get_parameters)
# delta - std diff
posterior(BF_ttest_paired, iterations = 4000) |> as.data.frame() |> 
  describe_posterior()
#> Summary of Posterior Distribution
#> 
#> Parameter | Median |         95% CI |     pd |          ROPE | % in ROPE
#> ------------------------------------------------------------------------
#> mu        |  -4.25 | [-6.83, -1.49] | 99.52% | [-0.10, 0.10] |        0%
#> sig2      |  15.50 | [ 5.54, 38.81] |   100% | [-0.10, 0.10] |        0%
#> delta     |  -1.09 | [-1.98, -0.26] | 99.52% | [-0.10, 0.10] |        0%
#> g         |   1.24 | [ 0.04, 18.28] |   100% | [-0.10, 0.10] |     0.66%
```

<sup>Created on 2021-12-21 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Reply by bwiernik** [reply]
Let's separate two points here:

1. `get_parameters()` returns the raw difference. I think this makes sense, at least as the default, as that is what is being tested in a t test. A `standardize` argument could be added. If the difference is standardized, the uncertainty in the SD should ideally be incorporated. 

2. BayesFactor uses analytic methods, not MCMC, so new posterior draws are computed by our methods, hence the MCMC instability on repeated calls

**Reply by mattansb** [reply]
> 1. `get_parameters()` returns the raw difference. I think this makes sense, at least as the default, as that is what is being tested in a t test. [...]

But is it? t-test is scale invariant, so I would argue that it actually tests a standardized difference!
Regardless, the priors (and parametrization) in `BayesFactor` are framed in terms of std-diffs, so it is at least worth showing both by default.

**Reply by bwiernik** [reply]
It's "scale invariant" only in that the pivot statistic for the mean is scale invariant. Testing an observed standardized difference means that uncertainty in the SD needs to be taken into account (eg, by its own pivot using the noncentrality parameter)

**Reply by mattansb** [reply]
<tangent 😜>
Hmmm... Isn't that exactly what a t (vs z) test does? Not sure I follow how t/p-values would / should be different between std and raw differences.

I was gonna try an simulate some data to see if the distribution of t values for raw vs standardized is different, but I can't even see how they can be different...

I know that for std-differences, CIs need to account for uncertainty in the standardizing factor, but if a NHST t-test is significant, than 0 is outside of the CI for the std-diff.

``` r
set.seed(1)
x <- rnorm(10)
y <- x - 0.7334084

t.test(x, y, var.equal = TRUE)[["p.value"]]
#> [1] 0.0499999

effectsize::cohens_d(x, y) |> print(digits = 7)
#> Cohen's d |                 95% CI
#> ----------------------------------
#> 0.9395614 | [0.0000004, 1.8559572]
#> 
#> - Estimated using pooled SD.
```

<sup>Created on 2021-12-24 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Reply by bwiernik** [reply]
Yes, I suppose for the hypothesis value of 0, the noncentral T distribution reduces to the central

**Reply by mattansb** [reply]
Oh no! I conflated "significant testing" with "*null hypothesis* significant testing"! mea culpa!

**Reply by mattansb** [reply]
Anyway, in `BayesFactor` the uncertainty in the SD is accounted for, so the test really is on the standardized scale.

**Comment by sloughon** [comment]
Thank both of you for fast and very helpful answers! Now I understand why the values ​​were different every time the code was run - because the samples are simulated and why the values ​​were different than expected - because I was building a ROPE interval based on raw differences thinking they were effect size values.

Courtesy of **mattansb**, I used the following code:

```
library(BayesFactor)
library(bayestestR)
a=c(value1, value2...)
b=c(value1, value2...)
result <- ttestBF(a, b)
result_posts <- posterior(result, iterations = 4e3)
d_posts <- result_posts[, "delta", drop = FALSE] 
rope(d_posts, range = c(-0.5, Inf),  ci = 1)
# Proportion of samples inside the ROPE [Inf, -0.50]:

Parameter | inside ROPE
-----------------------
delta     |     82.72 %
```

Now everything works and looks correct. I just wanted to make sure that I can interpret the above result as follows:

"there is 82.72% probability that effect size (in my case it was independent sample t-test so it is Cohen's d) is no lower then -0.5 (i.e., 82.72% of the posterior distribution falls within the ROPE) 

[I put my data in the sentence from Kruschke JK, Liddell TM. The Bayesian New Statistics: Hypothesis testing, estimation, meta-analysis, and power analysis from a Bayesian perspective. Psychon Bull Rev. 2018;25(1):178-206. doi:10.3758/s13423-016-1221-4]


Again, I will be grateful for any help.

**Reply by bwiernik** [reply]
That's reasonable (if you additionally add, "conditional on the data and my prior specification")

**Reply by sloughon** [reply]
Thanks for all the help

**Comment by sloughon** [comment]
After reading the above answers and a few attempts on my own analysis, I updated my question (see the post above) - thank you very much for your help and I would be grateful for a look also at it. I hope this post is not spam - delete if necessary.

Again, thank you in advance.
